cpp/include/cudf/detail/join.hpp

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cudf/column/column.hpp>
+#include <cudf/detail/structs/utilities.hpp>
+#include <cudf/detail/utilities/hash_functions.cuh>
+#include <cudf/table/table_view.hpp>
+#include <cudf/types.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/mr/device/polymorphic_allocator.hpp>
+
+#include <cuco/static_multimap.cuh>
+
+#include <cstddef>
+#include <memory>
+#include <optional>
+#include <utility>
+#include <vector>
+
+// Forward declaration
+template <typename T>
+class default_allocator;
+
+namespace cudf {
+namespace detail {
+
+constexpr int DEFAULT_JOIN_CG_SIZE = 2;
+
+enum class join_kind { INNER_JOIN, LEFT_JOIN, FULL_JOIN, LEFT_SEMI_JOIN, LEFT_ANTI_JOIN };
+
+/**
+ * @brief Hash join that builds hash table in creation and probes results in subsequent `*_join`
+ * member functions.
+ *
+ * User-defined hash function can be passed via the template parameter `Hasher`
+ *
+ * @tparam Hasher Unary callable type
+ */
+template <typename Hasher>
+struct hash_join {
```

Comment:
It feels like there's a lot of redundancy between `detail::hash_join` and `hash_join_impl`. 

Can we just merge them and make the PIMPL of the `cudf::hash_join` be a `unique_ptr<detail::hash_join<default_hash>>` ?

Reply:
That's what I tried in the first place. The issue is that I don't want to expose `detail::default_hash` in `cudf/join.hpp` since it's declared in `cudf/detail/utilities/hash_functions.cuh`. Including `hash_functions.cuh` in `cudf/join.hpp` will make the join public header a `.cuh` file as well.

Using `hash_join_impl` as a wrapper to hide `detail::default_hash` can minimize header exposure but like you pointed out, introduces redundancy. Please advise if there is a better solution.

Reply:
Fixed by forward-declaration of `MurmurHash3_32`